File: 01 Array/Product of Array Except Self/ProductExceptSelf.go
package main

import "fmt"

func productExceptSelf(nums []int) []int {
	n := len(nums)
	leftProduct := make([]int, n)
	rightProduct := make([]int, n)
	result := make([]int, n)
	for i := range leftProduct {
		leftProduct[i] = 1
		rightProduct[i] = 1
	}
	for i := 1; i < n; i++ {
		leftProduct[i] = leftProduct[i-1] * nums[i-1]
	}
	for i := n - 2; i >= 0; i-- {
		rightProduct[i] = rightProduct[i+1] * nums[i+1]
	}
	for i := 0; i < n; i++ {
		result[i] = leftProduct[i] * rightProduct[i]
	}
	return result
}

func main() {
	var t int
	fmt.Scan(&t)
	for t > 0 {
		t--
		var n int
		fmt.Scan(&n)
		nums := make([]int, n)
		for i := 0; i < n; i++ {
			fmt.Scan(&nums[i])
		}
		result := productExceptSelf(nums)
		for _, v := range result {
			fmt.Printf("%d ", v)
		}
		fmt.Println()
	}
}

File: 01 Array/Product of Array Except Self/ProductExceptSelf.java
import java.util.Arrays;
import java.util.Scanner;

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] leftProduct = new int[n];
        int[] rightProduct = new int[n];
        int[] result = new int[n];
        Arrays.fill(leftProduct, 1);
        Arrays.fill(rightProduct, 1);
        for (int i = 1; i < n; i++) {
            leftProduct[i] = leftProduct[i - 1] * nums[i - 1];
        }
        for (int i = n - 2; i >= 0; i--) {
            rightProduct[i] = rightProduct[i + 1] * nums[i + 1];
        }
        for (int i = 0; i < n; i++) {
            result[i] = leftProduct[i] * rightProduct[i];
        }
        return result;
    }
}

public class ProductExceptSelf {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0) {
            int n = sc.nextInt();
            int[] nums = new int[n];
            for (int i = 0; i < n; i++) {
                nums[i] = sc.nextInt();
            }
            Solution obj = new Solution();
            int[] result = obj.productExceptSelf(nums);
            for (int x : result) {
                System.out.printf("%d ", x);
            }
            System.out.println();
            t--;
        }
        sc.close();
    }
}